pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,470 @@
+	m.isIPv6Enabled, _ = config.IsIPv6Enabled(nodeConfig, trafficEncapMode)
+
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    m.onNodeAdd,
+		UpdateFunc: m.onNodeUpdate,
+		DeleteFunc: m.onNodeDelete,
+	})
+
+	nlmInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    m.onNodeLatencyMonitorAdd,
+		UpdateFunc: m.onNodeLatencyMonitorUpdate,
+		DeleteFunc: m.onNodeLatencyMonitorDelete,
+	})
+
+	return m
+}
+
+// Is current node
+func (m *NodeLatencyMonitor) isCurrentNode(node *corev1.Node) bool {
+	return node.Name == m.nodeName
+}
+
+// onNodeAdd is the event handler for adding Node.
+func (m *NodeLatencyMonitor) onNodeAdd(obj interface{}) {
+	node := obj.(*corev1.Node)
+	if m.isCurrentNode(node) {
+		return
+	}
+
+	m.latencyStore.addNode(node)
+
+	klog.V(4).InfoS("Node added", "Node", klog.KObj(node))
+}
+
+// onNodeUpdate is the event handler for updating Node.
+func (m *NodeLatencyMonitor) onNodeUpdate(oldObj, newObj interface{}) {
+	node := newObj.(*corev1.Node)
+	if m.isCurrentNode(node) {
+		return
+	}
+
+	m.latencyStore.updateNode(node)
+
+	klog.V(4).InfoS("Node updated", "Node", klog.KObj(node))
+}
+
+// onNodeDelete is the event handler for deleting Node.
+func (m *NodeLatencyMonitor) onNodeDelete(obj interface{}) {
+	node, ok := obj.(*corev1.Node)
+	if !ok {
+		deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.ErrorS(nil, "Received unexpected object", "obj", obj)
+			return
+		}
+		node, ok = deletedState.Obj.(*corev1.Node)
+		if m.isCurrentNode(node) {
+			return
+		}
+
+		if !ok {
+			klog.ErrorS(nil, "DeletedFinalStateUnknown contains non-Node object", "obj", deletedState.Obj)
+			return
+		}
+	}
+
+	m.latencyStore.deleteNode(node)
+}
+
+// onNodeLatencyMonitorAdd is the event handler for adding NodeLatencyMonitor.
+func (m *NodeLatencyMonitor) onNodeLatencyMonitorAdd(obj interface{}) {
+	nlm := obj.(*v1alpha1.NodeLatencyMonitor)
+	klog.V(4).InfoS("NodeLatencyMonitor added", "NodeLatencyMonitor", klog.KObj(nlm))
+
+	m.updateLatencyConfig(nlm)
+}
+
+// onNodeLatencyMonitorUpdate is the event handler for updating NodeLatencyMonitor.
+func (m *NodeLatencyMonitor) onNodeLatencyMonitorUpdate(oldObj, newObj interface{}) {
+	oldNLM := oldObj.(*v1alpha1.NodeLatencyMonitor)
+	newNLM := newObj.(*v1alpha1.NodeLatencyMonitor)
+	klog.V(4).InfoS("NodeLatencyMonitor updated", "NodeLatencyMonitor", klog.KObj(newNLM))
+
+	if oldNLM.GetGeneration() == newNLM.GetGeneration() {
+		return
+	}
+
+	m.updateLatencyConfig(newNLM)
+}
+
+// updateLatencyConfig updates the latency config based on the NodeLatencyMonitor CRD.
+func (m *NodeLatencyMonitor) updateLatencyConfig(nlm *v1alpha1.NodeLatencyMonitor) {
+	pingInterval := time.Duration(nlm.Spec.PingIntervalSeconds) * time.Second
+
+	m.latencyConfig = &LatencyConfig{
+		Enable:   true,
+		Interval: pingInterval,
+	}
+
+	m.latencyConfigChanged <- struct{}{}
```

Comment:
This is a blocking call and will wait for `monitorLoop` to process previous events. If this is desired, then having `m.latencyConfig` wouldn't make sense, as the channel can carry the config.

I guess the intention should be overriding the config, ensuring `monitorLoop` get notified and only realize the last config if multiple events come together. It should discard the notification if the channel is full:
```go
select {
        case m.latencyConfigChanged <- struct{}{}:
        default:
}
```

Reply:
I guess we could keep it as an unbuffered channel for simplicity and remove `m.latencyConfig `. This would also avoid the race condition you described in your comment above?

Reply:
I think it might be easier to just keep the unbuffered channel and send the latencyConfig directly.

Reply:
Sounds good to me, yes, it could avoid the race condition.